x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,550 @@
+        }
+
+        Builder storeMalformed(boolean storeMalformed) {
+            this.storeMalformed = storeMalformed;
+            return builder;
+        }
+
+        private VersionStringFieldType buildFieldType(BuilderContext context) {
+            boolean validateVersion = storeMalformed == false;
+            return new VersionStringFieldType(buildFullName(context), indexed, validateVersion, meta, boost, fieldType);
+        }
+
+        @Override
+        public VersionStringFieldMapper build(BuilderContext context) {
+            BooleanFieldMapper.Builder preReleaseSubfield = new BooleanFieldMapper.Builder(name + ".isPreRelease");
+            NumberType type = NumberType.INTEGER;
+            NumberFieldMapper.Builder majorVersionSubField = new NumberFieldMapper.Builder(name + ".major", type).nullValue(0);
+            NumberFieldMapper.Builder minorVersionSubField = new NumberFieldMapper.Builder(name + ".minor", type).nullValue(0);
+            NumberFieldMapper.Builder patchVersionSubField = new NumberFieldMapper.Builder(name + ".patch", type).nullValue(0);
+
+            return new VersionStringFieldMapper(
+                name,
+                fieldType,
+                buildFieldType(context),
+                storeMalformed,
+                nullValue,
+                multiFieldsBuilder.build(this, context),
+                copyTo,
+                preReleaseSubfield.build(context),
+                majorVersionSubField.build(context),
+                minorVersionSubField.build(context),
+                patchVersionSubField.build(context)
+            );
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        public TypeParser() {}
+
+        @Override
+        public Mapper.Builder<?> parse(String name, Map<String, Object> node, ParserContext parserContext) throws MapperParsingException {
+            Builder builder = new Builder(name);
+            TypeParsers.parseField(builder, name, node, parserContext);
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String propName = entry.getKey();
+                Object propNode = entry.getValue();
+                if (propName.equals("null_value")) {
+                    if (propNode == null) {
+                        throw new MapperParsingException("Property [null_value] cannot be null.");
+                    }
+                    builder.nullValue(propNode.toString());
+                    iterator.remove();
+                } else if (propName.equals("store_malformed")) {
+                    builder.storeMalformed(XContentMapValues.nodeBooleanValue(propNode, name + ".store_malformed"));
+                    iterator.remove();
+                } else if (TypeParsers.parseMultiField(builder::addMultiField, name, parserContext, propName, propNode)) {
```

Comment:
What's the need of  extra `TypeParsers.parseMultiField` on line 169 if `TypeParsers.parseField` on line 155 already takes care of multi-field?   Should we choose one of these 2 options?